R/get_mailchimp_data.R

```diff
@@ -27,91 +53,251 @@ get_mailchimp_data <- function(endpoint, date_since = NULL){
     "templates" = "since_created_at"
   )
 
-  access_api <- function(query, path){
-    query$offset <- 0
-    ret <- list()
-    while(TRUE){
-      url <- paste0("https://", area,".api.mailchimp.com/3.0/", path)
+  base_query <- list(count = 1000)
+  # set date filtering query depending on the endpoint
+  with_filter_query <- if(!is.null(date_since) &&
+                    !is.null(date_filter_params[[endpoint]])){
+    query <- base_query
+    query[[date_filter_params[[endpoint]]]] <- date_since
+    query
+  } else {
+    base_query
+  }
 
-      res <- httr::GET(url, httr::authenticate("any", api_key), query = query)
+  ret <- if (endpoint == "export/1.0/list") {
+    # get members list from export api
 
-      from_json <- res %>% httr::content(as = "text") %>% jsonlite::fromJSON()
+    # first, get these fields about lists from REST api
+    with_filter_query$fields <- paste0(
+      c(
+        "total_items",
+        "lists.id",
+        "lists.name",
+        "lists.date_created",
+        "lists.stats"
+      ), collapse = ",")
+    ret <- access_api(with_filter_query, dc, api_key, "lists")
 
-      # get last path to get the name of fetched objects
-      split <- stringr::str_split(path, "/")
-      if(length(split[[1]]) > 1) {
-        key <- tail(split[[1]], 1)
-      } else {
-        key <- path
-      }
+    if(length(ret) == 0){
+      stop("No data found.")
+    }
 
-      data <- tryCatch({
-        from_json[[key]] %>% jsonlite::flatten()
-      }, error = function(e){
-        NULL
-      })
+    # clean up data frame column names removing "stats." prefix
+    colnames(ret) <- stringr::str_replace(colnames(ret), "^stats\\.", "")
 
-      if(is.null(data) || length(data) == 0) {
-        break()
-      }
+    ret <- ret %>%
+      dplyr::select(-member_count, -merge_field_count, -last_sub_date, -last_unsub_date)
 
-      # somehow, offset query in /templates doesn't work,
-      # so check if access won't exceeds total number of the items
-      if(from_json$total_items >= query$offset + nrow(data)){
-        ret <- append(ret, list(data))
-        query$offset <- query$offset + query$count
-      } else {
-        break()
-      }
+    ids <- ret$id
+    # put "list_" suffix to colum names
+    # to make it clear that the columns are from list data
+    colnames(ret) <- paste0("list_", colnames(ret))
+
+    # get member data for each list from export API
+    ret$data <- lapply(ids, function(id){
+      export_members(id, dc, api_key, date_since)
+    })
+    ret %>%
+      unnest_without_empty(data)
```

Comment:
Why did we need this unnest_without_empty? Was there a lot of empty data? Please explain in the comment.